sql/src/main/java/io/crate/operation/projectors/ShardingUpsertExecutor.java

```diff
@@ -131,47 +141,130 @@ public ShardingUpsertExecutor(ClusterService clusterService,
+        this.consumeIteratorDelays = BACK_OFF_POLICY.iterator();
     }
 
     @Override
     public CompletableFuture<? extends Iterable<Row>> apply(BatchIterator batchIterator) {
-        CompletableFuture<List<Row>> result = new CompletableFuture<>();
-        consumeIterator(batchIterator).whenComplete((r, t) -> {
-            if (t == null) {
-                result.complete(Collections.singletonList(new Row1((long) r.cardinality())));
-            } else {
-                result.completeExceptionally(t);
-            }
-        });
+        safeConsumeIterator(batchIterator);
         return result;
     }
 
-    private CompletionStage<BitSet> consumeIterator(BatchIterator batchIterator) {
+    /**
+     * Consumes the iterator only if there isn't another thread already consuming it.
+     */
+    private void safeConsumeIterator(BatchIterator batchIterator) {
+        if (collectingEnabled.compareAndSet(false, true)) {
+            unsafeConsumeIterator(batchIterator);
+        }
+    }
+
+    /**
+     * Consumes the iterator without guarding against concurrent access to the iterator.
+     *
+     * !! THIS SHOULD ONLY BE CALLED BY THE loadNextBatch COMPLETE LISTENER !!
+     */
+    private void unsafeConsumeIterator(BatchIterator batchIterator) {
         Row row = RowBridging.toRow(batchIterator.rowData());
         try {
-            while (batchIterator.moveNext()) {
-                onRow(row);
-                if (idxWithinBatch == batchSize) {
-                    CompletableFuture<BitSet> executeBatchFuture = execute(false);
-                    idxWithinBatch = 0;
-                    return executeBatchFuture.thenCompose(r -> consumeIterator(batchIterator));
+            while (true) {
+                if (indexInBulk == bulkSize) {
+                    if (tryExecuteBulk(batchIterator, false) == false) {
+                        collectingEnabled.set(false);
+                        if (isScheduledCollectionRunning.compareAndSet(false, true)) {
+                            try {
+                                scheduleConsumeIterator(batchIterator);
+                            } catch (EsRejectedExecutionException e) {
+                                isScheduledCollectionRunning.set(false);
+                                if (collectingEnabled.compareAndSet(false, true)) {
+                                    // re-acquired the ownership of the iterator consumer, so let's try to
+                                    // execute / schedule this bulk again (otherwise someone else owns the
+                                    // iterator and carries on)
+                                    continue;
+                                }
+                            }
+                        }
+                        return;
+                    }
+                }
+
+                if (batchIterator.moveNext()) {
+                    onRow(row);
+                    pendingRequestsCount.incrementAndGet();
```

Comment:
`pendingRequestsCount` sounds like it contains the number of pending requests, but the counter here is incremented per row?